server/cluster.go

```diff
@@ -421,15 +441,81 @@ func (c *RaftCluster) RemoveStore(storeID uint64) error {
 		return errors.Trace(err)
 	}
 
+	// Remove an offline store should be OK, nothing to do.
+	if store.State == metapb.StoreState_Offline {
+		return nil
+	}
+
 	if store.State == metapb.StoreState_Tombstone {
 		return errors.New("store has been removed")
 	}
 
-	// TODO: Gracefully remove store.
+	store.State = metapb.StoreState_Offline
+	return c.saveStore(store)
+}
+
+// BuryStore marks a store as tombstone in cluster.
+// State transition:
+// Case 1: Up -> Tombstone (if force is true);
+// Case 2: Offline -> Tombstone.
+func (c *RaftCluster) BuryStore(storeID uint64, force bool) error {
+	c.Lock()
+	defer c.Unlock()
+
+	store, _, err := c.GetStore(storeID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Bury a tombstone store should be OK, nothing to do.
+	if store.State == metapb.StoreState_Tombstone {
+		return nil
+	}
+
+	if store.State == metapb.StoreState_Up {
+		if !force {
+			return errors.New("store is still up, please remove store gracefully")
+		}
+		log.Warnf("forcedly bury store %v", store)
+	}
+
 	store.State = metapb.StoreState_Tombstone
 	return c.saveStore(store)
 }
 
+func (c *RaftCluster) checkStores() {
+	cluster := c.cachedCluster
+	for _, store := range cluster.getMetaStores() {
+		if store.GetState() != metapb.StoreState_Offline {
+			continue
+		}
+		if cluster.regions.getStoreRegionCount(store.GetId()) == 0 {
+			err := c.BuryStore(store.GetId(), false)
```

Comment:
When the store is offline, bury it forcedly or not makes no difference, but we mean to do it gracefully here, so no force.